src/couch_replicator/README.md

```diff
@@ -0,0 +1,297 @@
+    restart `couch_multidb_changes` process it controls which will launch an
+    new scan of all the replicator databases.
+
+  * `couch_replicator_connection`: Maintains a global replication connection
+    pool. It allows reusing connection across replication tasks. Main interface
+    is a `acquire/1` and `release/1`. The main idea here is that once a
+    connection is established, it is kept around for
+    `replicator.connection_close_interval` milliseconds in case another
+    replication task wants to re-use it. It is worth pointing out how linking
+    and monitoring is handled: Workers are linked to the connection pool when
+    they are created. If they crash connection pool listens for the EXIT event
+    and cleans up. Connection pool also monitors owners (by monitoring the the
+    `Pid` from the `From` argument in the call to `acquire/1`) and cleans up if
+    owner dies. Another interesting thing is that connection establishment
+    (creation) happens in the owner process so the pool is not blocked on it.
+
+ * `couch_replicator_rate_limiter` : Implements a rate limiter to handle
+    connection throttling from sources or targets where requests return 429
+    error codes. Uses the Additive Increase / Multiplicative Decrease feedback
+    control algorithm to converge on the channel capacity. Implemented using a
+    16-way sharded ETS table to maintain connection state. The table sharding
+    code is split out to `couch_replicator_rate_limiter_tables` module. The
+    main idea of the module it so maintain and continually estimate an interval
+    for each connection represented by the `{Method, Url}`. The interval is
+    updated accordingly on each call to `failure/1` or `success/1` calls. A
+    `failure/1` is supposed to be called after a 429 is received and
+    `success/1` when a successful request has been made. Also when no failures
+    are happening the code is ensuring the ETS tables are empty in order to
+    have a lower impact on a running system.
+
+ * `couch_replicator_scheduler` : Scheduler is the core component of the
+    scheduling replicator. It allows handling a larger number of jobs than
+    might be possible to actively run on the cluster. It accomplishes this by
+    switching between jobs (stopping some and starting others) to ensure all
+    make progress. Replication jobs which fail are penalized using exponential
+    backoff. That is, each consecutive failure will double the time penalty.
+    This frees up system resources for more useful work than just continuously
+    trying to run the same subset of failing jobs.
+
+    The main API function is `add_job/1`. Its argument is an instance of
+    `#rep{}` record, which could also be the result of a document update from a
+    _replicator db or it could be the result of a POST to `_replicate`
+    endpoint. Once the replication job is added to the scheduler it doesn't
+    matter much where it originated.
```

Comment:
"matter much" because the behavior is different when replication jobs crash. 

Here is the relevant section:

```
% Handle crashed jobs. Handling differs between transient and permanent jobs.
% Transient jobs are those posted to the _replicate endpoint. They don't have a
% db associated with them. When those jobs crash, they are not restarted. That
% is also consistent with behavior when the node they run on, crashed and they
% do not migrate to other nodes. Permanent jobs are those created from
% replicator documents. Those jobs, once they pass basic validation and end up
% in the scheduler will be retried indefinitely (with appropriate exponential
% backoffs).
-spec handle_crashed_job(#job{}, any(), #state{}) -> ok.
handle_crashed_job(#job{rep = #rep{db_name = null}} = Job, Reason, State) ->
   ....
```

But in general most other behavior is the same